### PR TITLE
BACK-900 Save transactions in db

### PR DESCRIPTION
**Pull request by estrauser-ledger**
## What is this about?

- doobie request
- IT test PG

## Cute picture of lama
images

### CODE REVIEW

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Queries.scala [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import doobie.free.connection.ConnectionIO
+import doobie.implicits._
+import fs2.Stream
+
+object Queries {
+
+  def fetchTransaction(hash: String): Stream[ConnectionIO, Transaction] =
+    sql"""SELECT hash, account_id, block_hash, fee
+          FROM transaction
+          WHERE hash = $hash::bytea
```
Comment (onyb) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Queries.scala:13]: Why a byte array and not a string?

File: coins/bitcoin/interpreter/src/it/scala/co.ledger.lama.bitcoin.interpreter/TestResources.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    ce <- ExecutionContexts.fixedThreadPool[IO](conf.postgres.poolSize)
+    te <- ExecutionContexts.cachedThreadPool[IO]
+    db <- HikariTransactor.newHikariTransactor[IO](
+      conf.postgres.driver,
+      dbUrl,
+      dbUser,
+      dbPassword,
+      ce,
+      Blocker.liftExecutionContext(te)
+    )
```
Comment (jjayet-ledger) [inline coins/bitcoin/interpreter/src/it/scala/co.ledger.lama.bitcoin.interpreter/TestResources.scala:54]: You should replace that with:
`db <- postgresTransactor(conf.postgres)`

**Review by jjayet-ledger — CHANGES_REQUESTED** [review]

File: coins/bitcoin/interpreter/src/it/scala/co.ledger.lama.bitcoin.interpreter/TestResources.scala (hunk 2 above)
Comment (estrauser-ledger) [reply coins/bitcoin/interpreter/src/it/scala/co.ledger.lama.bitcoin.interpreter/TestResources.scala:54]: Oh yes :+1:

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Queries.scala (hunk 1 above)
Comment (estrauser-ledger) [reply coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Queries.scala:16]: It was Valentin's suggestion, it takes less space : https://github.com/LedgerHQ/lama/pull/15 (somewhere in Valentin's change suggestions)

File: build.sbt [hunk 3]
```diff
@@ -116,12 +121,12 @@ lazy val btcInterpreter = (project in file("coins/bitcoin/interpreter"))
...
     scalapbCodeGeneratorOptions += CodeGeneratorOption.FlatPackage,
     // Flyway credentials to migrate sql scripts
-    flywayLocations += "db",
-    flywayUrl := "jdbc:postgresql://localhost:5432/lama-btc-interpreter",
-    flywayUser := "lama",
-    flywayPassword := "serge"
+    flywayLocations += "db/migration",
+    flywayUrl := "jdbc:postgresql://localhost:5433/lama_btc",
+    flywayUser := "lama_btc",
+    flywayPassword := "serge_btc"
```
Comment (nukky94) [inline build.sbt:127]: For convenience, no need to have an other user, you can use the only and unique `serge lama`

File: coins/bitcoin/interpreter/src/it/resources/application.conf [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# SQL Database Configuration.
+postgres = {
+  url = "jdbc:postgresql://localhost:5433/lama_btc"
+  user = "lama_btc"
+  password = "serge_btc"
```
Comment (nukky94) [inline coins/bitcoin/interpreter/src/it/resources/application.conf:5]: lama serge

File: coins/bitcoin/interpreter/src/it/resources/application.conf [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# SQL Database Configuration.
+postgres = {
+  url = "jdbc:postgresql://localhost:5433/lama_btc"
```
Comment (nukky94) [inline coins/bitcoin/interpreter/src/it/resources/application.conf:3]: It's better to have a dedicated db`test_lama_btc` for IT tests

File: coins/bitcoin/interpreter/src/it/resources/db/migration/V999__populate_transactions.sql [hunk 6]
```diff
@@ -0,0 +1,57 @@
+INSERT INTO block VALUES (
```
Comment (nukky94) [inline coins/bitcoin/interpreter/src/it/resources/db/migration/V999__populate_transactions.sql:1]: why v999 ?

File: coins/bitcoin/interpreter/src/it/scala/co.ledger.lama.bitcoin.interpreter/QueriesIT.scala [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+  "account_id fetched from db" should "match account_id from migration" in IOAssertion {
+    setup() *>
+      appResources.use { db =>
+        fetchTx(db, transactionTest.accountId, transactionTest.hash)
+          .map { transaction =>
+            withClue("There should be at least one transaction in db.") {
+              transaction should not be None
+            }
+            transaction.map(_.accountId should be(transactionTest.accountId))
```
Comment (nukky94) [inline coins/bitcoin/interpreter/src/it/scala/co.ledger.lama.bitcoin.interpreter/QueriesIT.scala:21]: `transaction.map(_.accountId) should contain only transactionTest.accountId`

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Queries.scala [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+          """
+      .query[(String, String, String, Int)]
+      .map {
+        case (account_id, hash, block_hash, fee) => Transaction(account_id, hash, block_hash, fee)
+      }
+      .stream
+
+  def fetchTransactionForAccount(
```
Comment (nukky94) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Queries.scala:26]: `fetchTransactions`

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Queries.scala [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+import cats.implicits._
+import doobie.implicits._
+import doobie._
+import doobie.implicits.javasql._
+
+object Queries {
+
+  implicit val bigIntType: Meta[BigInt] = Meta.Advanced.other[BigInt]("bigint")
+
+  def fetchTransaction(hash: String): Stream[ConnectionIO, Transaction] =
```
Comment (nukky94) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Queries.scala:15]: `ConnectionIO[Option[Transaction]]`

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Queries.scala [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    sql"""SELECT account_id, hash, block_hash, fee
+          FROM transaction
+          WHERE hash = $hash::bytea
+          """
+      .query[(String, String, String, Int)]
+      .map {
+        case (account_id, hash, block_hash, fee) => Transaction(account_id, hash, block_hash, fee)
+      }
+      .stream
```
Comment (nukky94) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Queries.scala:24]: .option

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Queries.scala [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+          WHERE hash = $hash::bytea
+          AND account_id = $accountId
+          """
+      .query[(String, String, String, Int)]
+      .map {
+        case (account_id, hash, block_hash, fee) => Transaction(account_id, hash, block_hash, fee)
+      }
+      .stream
+
+  def saveTransaction(tx: Transaction): ConnectionIO[Int] =
```
Comment (nukky94) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Queries.scala:41]: To optimize later regarding batch updates:
https://tpolecat.github.io/doobie/docs/07-Updating.html#batch-updates

**Review by nukky94 — CHANGES_REQUESTED** [review]

File: coins/bitcoin/interpreter/src/it/resources/db/migration/V999__populate_transactions.sql (hunk 6 above)
Comment (estrauser-ledger) [reply coins/bitcoin/interpreter/src/it/resources/db/migration/V999__populate_transactions.sql:1]: It's a test migration but the names conflict with the main migrations and there's already a V1. I could name it V2 , but when there'll be a V2 in the main migrations, we'll have to rename the test migration to be V3, and so on.
So I went with 999 (wich is OVERKILL!) but we could name it whatever, what's your suggestion ?

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Queries.scala (hunk 8 above)
Comment (estrauser-ledger) [reply coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Queries.scala:26]: I'll also remove the fetchTransaction which does not have the accountId parameter, it doesn't make any sense.

File: coins/bitcoin/interpreter/src/it/resources/db/migration/V999__populate_transactions.sql (hunk 6 above)
Comment (nukky94) [reply coins/bitcoin/interpreter/src/it/resources/db/migration/V999__populate_transactions.sql:1]: Hmm, maybe it's the wrong way to use migration in order to insert data for testing :D

**Review by nukky94 — APPROVED** [review]

**Pull request merged**